test/tests/layer.js

```diff
@@ -82,7 +89,30 @@ suite("d3.layer", function() {
 			this.layer.draw([]);
 			assert(this.insert.calledOn(this.dataBind.returnValues[0].enter.returnValues[0]));
 		});
-		
+		test("by default removes exiting nodes from the DOM", function() {
+			this.layer.draw([1]);
+			assert.equal(this.layer.selectAll("g").size(), 1);
+			this.layer.draw([]);
+			assert.equal(this.layer.selectAll("g").size(), 0);
+		});
+		test("invokes the provided `remove` method in the context of the layer's bound 'exit' selection", function() {
+			var updating, exiting;
+
+			this.layerWithRemove.draw([1, 2, 3]);
+			this.layerWithRemove.draw([]);
+
+			updating = this.dataBind.returnValues[1];
+			exiting = updating.exit.returnValues[0];
+
+			assert(this.remove.calledOn(exiting));
```

Comment:
So, I think it's premature to assume that drawing an empty set is equivalent to removing the layer. I can see, for example, visualizing streaming data might result in gaps that would just not have any nodes in the rendering cycle for that subset, but that might change come next data frame. I think this also adds a bit of unstated "magic" whereby we're saying calling `draw` with an empty set will result in the layer removal being called. I would prefer that were a more explicit user choice.